pkg/metrics/metrics.go

```diff
@@ -69,18 +109,39 @@ type MetricsManager interface {
 	// op - the operation which the metric is associated with.
 	// status - the operation status, if not specified, i.e., status == nil, an
 	//          "Unknown" status of the passed-in operation is assumed.
-	RecordMetrics(op Operation, status OperationStatus)
+	RecordMetrics(op Operation, status OperationStatus, volumeSnapshot *crdv1.VolumeSnapshot)
 }
 
 // Operation is a structure which holds information to identify a snapshot
 // related operation
 type Operation struct {
-	// the name of the operation, for example: "CreateSnapshot", "DeleteSnapshot"
+	// Name is the name of the operation, for example: "CreateSnapshot", "DeleteSnapshot"
 	Name string
-	// the name of the driver which executes the operation
+	// Driver is the driver name which executes the operation
 	Driver string
```

Comment:
discussed offline.
Putting driver name into the key (operation) seems to bring quite some problems with little benefits.
The name of the operation + resource UID should already be unique enough.
same thing applies to snapshot type.
There is no guarantee that a valid VolumeSnapshotClass exists when a VolumeSnapshot CR is created for dynamic provisioning.

given that, I am proposing the following:

key ->
type Operation struct {
  operation name
  resource uid
}

value:
type val struct {
  start ts
  driver
  snapshot type
}